### PR TITLE
LPD-33602 Technical Task | Delete document metadata sets via headless API

### PR DESCRIPTION
**Pull request by SamZiemer**
# What is this trying to solve?

https://liferay.atlassian.net/browse/LPD-33602

Create a new endpoint for Headless API to delete DocumentMetadataSets

# How am I fixing it?

Added a new endpoint for the delete method

# How can you verify that it works?

Run the tests in DocumentMetadataSetResourceTest

Resending this PR because Brian Chan suggested using the local service instead of DDMStructureService.
https://github.com/brianchandotcom/liferay-portal/pull/154100#issuecomment-2344416825

### CODE REVIEW

**Comment by SamZiemer** [comment]
ci:test:relevant

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/DocumentMetadataSetResourceImpl.java [hunk 1]
```diff
@@ -40,6 +41,20 @@
...
+	@Override
+	public void deleteDocumentMetadataSet(Long documentMetadataSetId)
+		throws Exception {
+
+		if (!FeatureFlagManagerUtil.isEnabled("LPD-32247")) {
+			throw new UnsupportedOperationException();
+		}
+
+		_ddmStructureLinkLocalService.deleteStructureStructureLinks(
+			documentMetadataSetId);
```
Comment (adolfopa) [inline modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/DocumentMetadataSetResourceImpl.java:53]: What will happen if there are documents that contain values for this metadata set? Can you verify that the document is still visible, and that it doesn't try to look for the structure?

Also, shouldn't you clean up the DLFileEntryMetadata entries as well then?

**Review by adolfopa — CHANGES_REQUESTED** [review]

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/DocumentMetadataSetResourceImpl.java (hunk 1 above)
Comment (SamZiemer) [reply modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/DocumentMetadataSetResourceImpl.java:55]: Hi @adolfopa, you are correct, we should be removing the DLFileEntryMetadata entries as well. I've pushed another commit to do that. I do have a question about it though, currently we do not have a method to delete DLFileEntryMetadata by structureId, so using the methods available now, I am able to delete the structure, then use _dlFileEntryMetadataLocalService.getNoStructuresFileEntryMetadatas() do retrieve a list of entries that no longer have a structure associated with them.

My question is should this be the approach, or should I add a new method in the service layer to delete by structureId?

> What will happen if there are documents that contain values for this metadata set? Can you verify that the document is still visible, and that it doesn't try to look for the structure?
Regarding this question, yes we can still view the document. The subsection in the information tab for the metadata set is no longer present and no errors are thrown.

**Comment by SamZiemer** [comment]
ci:test:sf

**Comment by SamZiemer** [comment]
ci:test:relevant

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/DocumentMetadataSetResourceImpl.java (hunk 1 above)
Comment (adolfopa) [reply modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/DocumentMetadataSetResourceImpl.java:55]: I think we should add a specific method to the Java API. It should be a matter of simply adding a new finder in the service.xml file.

**Comment by SamZiemer** [comment]
ci:test:sf

**Comment by SamZiemer** [comment]
ci:test:relevant

File: modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/DocumentMetadataSetResourceImpl.java (hunk 1 above)
Comment (SamZiemer) [reply modules/apps/headless/headless-delivery/headless-delivery-impl/src/main/java/com/liferay/headless/delivery/internal/resource/v1_0/DocumentMetadataSetResourceImpl.java:54]: Thanks Adolfo, I've pushed some additional commits to make that change.

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryMetadataLocalServiceImpl.java [hunk 2]
```diff
@@ -60,6 +60,18 @@ public void deleteFileEntryMetadata(long fileEntryId)
...
+	@Override
+	public void deleteFileEntryMetadataByDDMStructureId(long ddmStructureId)
+		throws PortalException {
+
+		for (DLFileEntryMetadata dlFileEntryMetadata :
+				dlFileEntryMetadataPersistence.findByDDMStructureId(
+					ddmStructureId)) {
+
+			deleteFileEntryMetadata(dlFileEntryMetadata);
+		}
```
Comment (adolfopa) [inline portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryMetadataLocalServiceImpl.java:72]: Why not use `removeByDDMStructureId` here?

**Review by adolfopa — CHANGES_REQUESTED** [review]

File: portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryMetadataLocalServiceImpl.java (hunk 2 above)
Comment (SamZiemer) [reply portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryMetadataLocalServiceImpl.java:72]: [deleteFileEntryMetadata](https://github.com/SamZiemer/liferay-portal/blob/master/portal-impl/src/com/liferay/portlet/documentlibrary/service/impl/DLFileEntryMetadataLocalServiceImpl.java#L31-L47) also removes DDM storage and structure links associated with the fileEntryMetadataId, which I was not catching in the 		_ddmStructureLinkLocalService.deleteStructureStructureLinks(documentMetadataSetId) call.

**Comment by SamZiemer** [comment]
ci:test:relevant

**Comment by SamZiemer** [comment]
ci:test:relevant

**Review by adolfopa — APPROVED** [review]

**Comment by adolfopa** [comment]
ci:forward

**Comment by SamZiemer** [comment]
ci:reevaluate:1317401_11407

**Comment by ambrinchaudhary** [comment]
ci:report:14586452

**Comment by SamZiemer** [comment]
ci:forward:force

**Pull request closed without merging**